### ISSUE TITLE
QAT callback doesn't work with multi-input models or forward kwargs

### ISSUE THREAD

**Issue by addisonklinke** [Lightning-AI/lightning#7926]
## 🐛 Bug

The `QuantizationAwareTraining` can only handle models whose forward signature looks like `forward(data)`. Adding additional positional or keyword arguments breaks the forward propagation.

## Please reproduce using the BoringModel

The `BoringModel` doesn't have the right forward signature to produce errors, but I included an equivalently simple model below

### To Reproduce

```python
import functools
import torch
import torch.nn as nn


class Simple(nn.Module):
    def __init__(self):
        super(Simple, self).__init__()
        self.conv = nn.Conv2d(3, 16, kernel_size=3)
        self.softmax = nn.Softmax(dim=3)

    def forward(self, a, b, softmax=False):
        """Use 2 positional arguments and 1 keyword argument"""
        feat_b = self.conv(b)
        feat_a = self.conv(a)
        combined = feat_a + feat_b
        if softmax:
            return self.softmax(combined)
        else:
            return combined

    @property
    def example_input_array(self):
        return torch.randn(1, 3, 28, 28), torch.randn(1, 3, 28, 28)


def wrap_quantize_forward_context(model, func):
    """From QuantizationAwareTraining callback"""
    @functools.wraps(func)
    def wrapper(data):
        data = model.quant(data)
        data = func(data)
        data = model.dequant(data)
        return data
    return wrapper


# Works fine without quantization
model = Simple()
input_sample = model.example_input_array
_ = model(*input_sample)
_ = model(*input_sample, softmax=True)

# Simulate the `on_fit_start` hook behavior
model.quant = torch.quantization.QuantStub()
model.dequant = torch.quantization.DeQuantStub()
model.forward = wrap_quantize_forward_context(model, model.forward)
_ = model(*input_sample)
_ = model(*input_sample, softmax=True)
```

The last two lines will raise the following exceptions

```
TypeError: wrapper() takes 1 positional argument but 2 were given
TypeError: wrapper() got an unexpected keyword argument 'softmax'
```

### Expected behavior

Forward prop should work fine for both the quantized and unquantized model

### Environment

* CUDA:
        - GPU: GeForce GTX 1060 with Max-Q Design
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1+cu102
        - pytorch-lightning: 1.3.0
        - tqdm:              4.59.0
* System:
        - OS:                Linux
        - architecture: 64bit
        - processor:         x86_64
        - python:            3.6.13
        - version:           #148-Ubuntu SMP Sat May 8 02:33:43 UTC 2021

### Additional context

The solution to the minimal example is to make the decoration wrapper more generic

```python
def wrap_quantize_forward_context(model, func):
    """From QuantizationAwareTraining callback"""
    @functools.wraps(func)
    def wrapper(*data, **kwargs):
        data = model.quant(data)
        data = func(*data, **kwargs)
        data = model.dequant(data)
        return data
    return wrapper
```

However, I notice that `callbacks/quantization.py` also includes a second decorator `wrap_qat_forward_context`, and I am not exactly clear on the intent behind having two separate decorators. I see that `wrap_quantize_forward_context` is only used in the `on_fit_end` hook. Any insight here @Borda?

I can make a PR for the above solution, but first I tested it in an actual training by patching my local install

```python
def wrap_qat_forward_context(
    quant_cb,
    model: 'pl.LightningModule',
    func: Callable,
    trigger_condition: Optional[Union[Callable, int]] = None
) -> Callable:
    @functools.wraps(func)
    def wrapper(*data, **kwargs) -> Any:  # <-- Changes to decoration signature

        _is_func_true = isinstance(trigger_condition, Callable) and trigger_condition(model.trainer)
        _is_count_true = isinstance(trigger_condition, int) and quant_cb._forward_calls < trigger_condition
        _quant_run = trigger_condition is None or _is_func_true or _is_count_true
        if _quant_run:
            quant_cb._forward_calls += 1
            data = model.quant(data)  # <-- Triggers AttributeError 'tuple' object has no attribute 'detach'
        data = func(*data, **kwargs)
        if _quant_run:
            data = model.dequant(data)
        return data

    return wrapper
```

So I tried also expanding the `data` before passing to `model.quant` but this goes back to a `TypeError`

```python
def wrap_qat_forward_context(
    quant_cb,
    model: 'pl.LightningModule',
    func: Callable,
    trigger_condition: Optional[Union[Callable, int]] = None
) -> Callable:
    @functools.wraps(func)
    def wrapper(*data, **kwargs) -> Any:

        _is_func_true = isinstance(trigger_condition, Callable) and trigger_condition(model.trainer)
        _is_count_true = isinstance(trigger_condition, int) and quant_cb._forward_calls < trigger_condition
        _quant_run = trigger_condition is None or _is_func_true or _is_count_true
        if _quant_run:
            quant_cb._forward_calls += 1
            data = model.quant(*data)  # <-- Attempted tuple unpacking, get TypeError: forward() takes 2 positional arguments but 3 were given
        data = func(*data, **kwargs)
        if _quant_run:
            data = model.dequant(*data)
        return data

    return wrapper
```

**Comment by addisonklinke** [comment]
After some more research, I found a [post](https://discuss.pytorch.org/t/115018/2) on the PyTorch forums that shows how to setup quant stubs for a multi-input model. Apparently you need to have a separate stub for each tensor. That means the `on_fit_start` method of the QAT callback will need to use some sort of introspection (or perhaps `model.example_input_array`) to determine how many quant stub attributes to create. Then the wrappers need to pair each stub with the incoming tensor

**Comment by Borda** [comment]
@addisonklinke do you suggest integrating it in PL?

**Comment by addisonklinke** [comment]
Yes, I think it would be reasonable to support multi-inputs and forward kwargs. Many of the models I work with require at least the former. I've been playing around locally with something like this

```python
def on_fit_start(self, trainer, pl_module):
    if pl_module.example_input_array is None:
        raise MisconfigurationException('Example input needed to determine number of QuantSubs for QAT')
    pl_module.quants = [torch.quantization.QuantStub() for _ in range(len(pl_module.example_input_array))]

@functools.wraps(func)
def wrapper(*data, **kwargs) -> Any:
    data = [q(d) for q, d in zip(model.quants, data)]
    data = func(*data, **kwargs)
    data = model.dequant(data)
    return data
```

I'll start a draft PR soon once I have it more finalized.

Even with the above changes, there is still an error after the fit completes. If I try running a forward pass, the `quants` attribute is no longer quantizing the input tensors, so then I wind up passing a CPU tensor to a quantized kernel which raises the runtime error described in the PyTorch [docs](https://pytorch.org/docs/stable/quantization.html#passing-a-non-quantized-tensor-into-a-quantized-kernel). Could you provide a more detailed explanation of the `input_compatible` flag? I think that might help me sort out the expected behavior post training - currently I wonder if I'm running inference incorrectly and that the runtime error could be avoided if I took a different approach

**Comment by tchaton** [comment]
Dear @addisonklinke,

Any progress on the draft PR ?

Best,
T.C

**Comment by addisonklinke** [comment]
Hoping to have it up today or tomorrow

**Comment by awaelchli** [comment]
Closing this stale issue because we removed the QAT callback in #16750. As explained in the linked PR:

The QAT callback can no longer be maintained by us. It has many issues that make the callback uneffective and these can't be fixed at the moment.

Users who rely on this callback: You can stay on the Lightning 1.9.x version which gets long-term support (LTS) OR you can copy the callback code and maintain it yourself.

If someone from the community is interested in fixing and maintaining this callback, please let us know. If important issues come up that affect several users, we will prioritize fixes accordingly for the 1.9.x LTS.
